quark.c

```diff
@@ -1996,6 +2590,57 @@ quark_queue_get_snap_event(struct quark_queue *qq)
 	return (qev);
 }
 
+static const struct quark_event *
+raw_event_connection(struct quark_queue *qq, struct raw_event *raw)
+{
+	struct quark_event	*qev;
+	struct quark_socket	*qsk;
+	struct raw_connection	*conn;
+
+	conn = &raw->connection;
+
+	qev = &qq->event_storage;
+
+	qsk = socket_cache_get(qq, &conn->local, &conn->remote, 1);
+	if (qsk == NULL)
+		return (NULL);
+
+	qsk->pid_last_use = raw->pid;
+
+	switch (conn->conn) {
+	case CONNECTION_ACCEPT:	/* FALLTHROUGH */
+	case CONNECTION_CONNECT:
+		/*
+		 * We might race, this socket might have been created via /proc
+		 * parsing, but we sill got an ACCEPT/CONNECT since the rings
+		 * were opened before
+		 */
+		if (qsk->pid_origin != 0 && qsk->established_time != 0)
+			return (NULL);
+		if (qsk->pid_origin == 0)
+			qsk->pid_origin = qsk->pid_last_use = raw->pid;
+		if (qsk->established_time == 0)
+			qsk->established_time = raw->time;
+		qev->events = QUARK_EV_CONNECTION_ESTABLISHED;
+		break;
+	case CONNECTION_CLOSE:
+		if (qsk->close_time == 0)
+			qsk->close_time = raw->time;
+		gc_mark(qq, &qsk->gc, GC_SOCKET);
+		qev->events = QUARK_EV_CONNECTION_CLOSED;
+
+		break;
+	default:
+		qwarnx("invalid conn->conn %d\n", conn->conn);
+		return (NULL);
```

Comment:
if `qsk` was allocated by `socket_cache_get`, should it be freed here?

Reply:
Yep, this is more like a programming error, but we should be correct anyway, I'll have to change this function a bit, because there is a second issue:
Basically socket_cache_get() cannot do the "lookup or alloc", it needs to be split into two. When we get an ACCEPT/CONNECT, we must hit the alloc case, which might incur us evicting the old entry and creating a new one.
This can be the case if, for any reason, we missed the close event.

Reply:
In the upcoming version, it doesn't have qsk allocated on that path anymore, so this is gone.